doc: hpf and softperipherals comparison (#23226)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/coprocessors/ug_hpf_softperipherals_comparison.rst

﻿.. _ug_hpf_softperipherals_comparison:

Introduction to High-Performance Framework and Soft Peripherals
###############################################################

.. contents::
   :local:
   :depth: 2

:ref:`nrfxlib:soft_peripherals` and the :ref:`High-Performance Framework (HPF)<hpf_index>` both facilitate the creation of software-based peripherals.

> Choosing between these solutions depends on the specific requirements of your project and the resources available.

Reviewer: Missing definition of what these two are. I feel an introduction should state that. You only state what their purpose is.

Reviewer: Comment not applied. @masz-nordic , @lopeztel , please assist @annwoj and provide proper definitions.

.. list-table::
   :header-rows: 1

   * - Feature
     - Soft Peripherals
     - High Performance Framework
   * - Integration
     - Pre-compiled binary for FLPR, driver for Application Core
     - Built from source in Zephyr build system
   * - Memory requirements
     - Fixed (~16K)
     - Adjustable, depends on code and devicetree
   * - Application Compatibility
     - Can be used on baremetal or Zephyr applications
     - Uses Zephyr without kernel on FLPR side, full Zephyr on APP side
   * - Integration complexity
     - Minor integration required
     - Samples provided for the start of development
   * - IPC mechanism
     - Register-based
     - ic(b)msg or mbox-based
   * - Compliance & testing
     - Protocol-compliant, verified by Nordic Semiconductor
     - Samples provided as-is, testing is done by the user
   * - Delivery form
     - Binary
     - Source code, modifiable and extendable
   * - API level
     - Exposed at HW-driver level
     - FLPR may handle parts of SW stack above HW-driver level

.. _nrf54l_hpf_softperi_comparison_creating_peripherals:

Creating software-based peripherals

> ***********************************

Reviewer: Gerund is commonly used in our docs to indicate user-actionable step procedure. This conceptual documentation does not provide such steps, so this should be rephrased not to suggest them.

For Soft Peripherals, the integration process primarily involves incorporating glue code to facilitate their use.

> To learn how to utilize those peripherals, and to see what is currently supported, refer to the :ref:`nrfxlib:soft_peripherals` documentation.

Reviewer: utilize > use

In contrast, HPF provides initial samples that serve as a starting point for development.
For detailed guide on creating your own custom peripherals, see the :ref:`hpf_index` page.

.. _nrf54l_hpf_softperi_comparison_limitations:

> Performance and Limitations

Reviewer: ```suggestion
Performance and limitations
```
We follow sentence style for headers.

They can be used in situations where hardware peripherals are absent or insufficient.

HPF, on the other hand, enables the development of custom, software-defined peripherals.
It enhances the main processor's functionality with real-time capabilities in a streamlined, baremetal environment, integrating with the Zephyr build system.
HPF can also offload parts of the protocol stack to the FLPR core.

.. _nrf54l_hpf_softperi_comparison_use_case:

Implementation and use cases
============================

The following comparison details implementation and typical use cases for Soft Peripherals and High Performance Framework:

.. list-table::
   :header-rows: 1

   * - Aspect
     - Soft Peripherals
     - High Performance Framework
   * - Development
     - Developed by Nordic Semiconductor, optimized for performance and power consumption.
     - Allows for the creation of specialized or custom peripherals.
   * - Use cases
     - When an application requires an extra or absent peripheral and a corresponding Soft Peripheral is available.

>      - When custom peripheral development or a high-level control is needed.

Reviewer 2: ```suggestion
     - When developing a custom peripheral or high degree of control is needed.
```

﻿.. _ug_hpf_softperipherals_comparison:

Introduction to High-Performance Framework and Soft Peripherals
###############################################################

.. contents::
   :local:
   :depth: 2

The :ref:`High-Performance Framework (HPF)<hpf_index>` is a framework designed to support the development and integration of software peripherals using coprocessors.
It offers tools such as Hardware Abstraction Layers (HALs) and CMake targets, along with application samples and guidelines on architectural issues like event, fault, and power management.
These resources help developers create customized software peripherals tailored to specific application needs.

> :ref:`nrfxlib:soft_peripherals`, on the other hand, are a collection of pre-compiled binaries and API driver code designed to emulate commonly used peripherals.

Reviewer 3: ```suggestion
:ref:`nrfxlib:soft_peripherals` are a collection of pre-compiled binaries and API driver code designed to emulate commonly used peripherals.
```

These peripherals allow application code to use the API driver to load the pre-compiled binaries onto a specific RAM location, where they are executed by the RISC-V coprocessor, named Fast Lightweight Peripheral Processor (FLPR).
The application can then control the soft peripherals using API functions.

Both methods facilitate the creation of software-based peripherals, that emulate the functionality of physical hardware peripherals (IP) through software.
Choosing between the solutions depends on the specific requirements of your project and the resources available.

Purpose
*******

Soft peripherals can be used in situations where hardware peripherals are absent or insufficient.

> HPF, on the other hand, enables the development of custom, software-defined peripherals.

Reviewer 3: ```suggestion
HPF enables the development of custom, software-defined peripherals.
```

     - When developing a custom peripheral or when a high-degree control is needed.
   * - Suitability
     - Suitable for applications built from pre-compiled sources.
     - Ideal for applications where the FLPR handles higher layers of the software stack.

.. _nrf54l_hpf_softperi_comparison_requirements:

Requirements
************

For Soft Peripherals, there is a specific memory size requirement of approximately 16K, but there is a flexibility in placement within the FLPR execution RAM.
To enter low power consumption modes, Soft Peripherals require a slot at a specific memory address, which is platform-specific, and the application code must be able to access MEMCONF registers.
You can check these requirements in the :ref:`Soft Peripherals section<nrfxlib:soft_peripherals>`, which provide essential information for proper integration and optimization.

On the other hand, HPF necessitates that the application core uses the Zephyr operating system.
The framework allows for a high degree of configurability, which you can tailor according to your needs using Kconfig and devicetree settings.

> This flexibility allows you to optimize the memory footprint and functionality of the custom peripherals, ensuring that the peripherals are well-integrated and perform efficiently within their specific application environments.

Reviewer 3: ```suggestion
This allows you to optimize the memory footprint and functionality of the custom peripherals, ensuring that the peripherals are well-integrated and perform efficiently within their specific application environments.
```

.. _nrf54l_hpf_softperi_comparison_features:

Features comparison
*******************

See the following detailed feature comparison between Soft Peripherals and High-Performance Framework:

.. list-table::
   :header-rows: 1

   * - Feature
     - Soft Peripherals
     - High Performance Framework
   * - Integration

>      - Pre-compiled binary for FLPR, driver for Application Core

Reviewer 3: ```suggestion
     - Pre-compiled binary for FLPR, driver for application core
```

     - Built from source in Zephyr build system
   * - Memory requirements

>      - Fixed (~16K)

Reviewer 3: ```suggestion
     - Fixed (~16 K)
```

     - Adjustable, depends on code and devicetree

>    * - Application Compatibility

Reviewer 3: ```suggestion
   * - Application compatibility
```

     - Can be used on baremetal or Zephyr applications
     - Uses Zephyr without kernel on FLPR side, full Zephyr on APP side
   * - Integration complexity
     - Minor integration required
     - Samples provided for the start of development
   * - IPC mechanism
     - Register-based
     - ic(b)msg or mbox-based

>    * - Compliance & testing

Reviewer 3: ```suggestion
   * - Compliance and testing
```

     - Protocol-compliant, verified by Nordic Semiconductor
     - Samples provided as-is, testing is done by the user
   * - Delivery form
     - Binary
     - Source code, modifiable and extendable
   * - API level

>      - Exposed at HW-driver level

Reviewer 3: ```suggestion
     - Exposed at hardware driver level
```

>      - FLPR may handle parts of SW stack above HW-driver level

Reviewer 3: ```suggestion
     - FLPR may handle parts of the software stack above hardware driver level
```

.. _nrf54l_hpf_softperi_comparison_creating_peripherals:

Creation of software-based peripherals
**************************************

For Soft Peripherals, the integration process primarily involves incorporating glue code to facilitate their use.
To learn how to use those peripherals, and to see what is currently supported, refer to the :ref:`nrfxlib:soft_peripherals` documentation.

In contrast, HPF provides initial samples that serve as a starting point for development.

> For detailed guide on creating your own custom peripherals, see the :ref:`hpf_index` page.

Reviewer 3: ```suggestion
For a detailed guide for creating your own custom peripherals, see the :ref:`hpf_index` page.
```

Author: leaving "on" to avoid repetitions

.. _nrf54l_hpf_softperi_comparison_limitations:

Performance and limitations
***************************

Each Soft Peripheral is unique and comes with its own set of limitations compared to traditional hardware IP.
These limitations are specific to the functions that the peripheral is designed to emulate and how they integrate with the rest of the system.

> HPF, while offering extensive customization and control, is currently in an experimental stage and lacks full Power Management support.

Reviewer 3: Do we need caps for power management?

﻿.. _ug_hpf_softperipherals_comparison:

Introduction to High-Performance Framework and Soft Peripherals
###############################################################

.. contents::
   :local:
   :depth: 2

The :ref:`High-Performance Framework (HPF)<hpf_index>` is a framework designed to support the development and integration of software peripherals using coprocessors.
It offers tools such as Hardware Abstraction Layers (HALs) and CMake targets, along with application samples and guidelines on architectural issues like event, fault, and power management.
These resources help developers create customized software peripherals tailored to specific application needs.

:ref:`nrfxlib:soft_peripherals` are a collection of pre-compiled binaries and API driver code designed to emulate commonly used peripherals.

> These peripherals allow application code to use the API driver to load the pre-compiled binaries onto a specific RAM location, where they are executed by the RISC-V coprocessor, named Fast Lightweight Peripheral Processor (FLPR).

Reviewer: ```suggestion
These peripherals allow application code to use the API driver to load the pre-compiled binaries onto a specific RAM location, where they are executed by :ref:`Fast Lightweight Peripheral Processor (FLPR) <vpr_flpr_nrf54l>`, a RISC-V coprocessor.
```

Author: I wouldn't add this - didn't want to reference the nRF54L docs considering sQSPI supports both H20 and L15

Reviewer: I would then add mention of the hardware platform differences here.

The application can then control the soft peripherals using API functions.

> Both methods facilitate the creation of software-based peripherals that emulate the functionality of physical hardware peripherals (IP) through software.
> Choosing between the solutions depends on the specific requirements of your project and the resources available.
> 
> Purpose
> *******

Reviewer: ```suggestion
Purpose
*******

Both methods facilitate the creation of software-based peripherals that emulate the functionality of physical hardware peripherals (IP) through software.
Choosing between the solutions depends on the specific requirements of your project and the resources available.
```
Maybe "Design rationale and usage" instead of "Purpose"?

> You can use soft peripherals when hardware peripherals are absent or insufficient.
> 
> HPF enables the development of custom, software-defined peripherals.
> It enhances the main processor's functionality with real-time capabilities in a streamlined, baremetal environment, integrating with the Zephyr build system.
> HPF can also offload parts of the protocol stack to the FLPR core.

Reviewer: ```suggestion
* You can use soft peripherals when hardware peripherals are absent or insufficient.

* You can use HPF to enable the development of custom, software-defined peripherals.
  It enhances the main processor's functionality with real-time capabilities in a streamlined, bare metal environment, integrating with the Zephyr build system.
  HPF can also offload parts of the protocol stack to the FLPR core.
```

Author: reworded line 26 to avoid repetitions

.. _nrf54l_hpf_softperi_comparison_use_case:

Implementation and use cases
============================

The following comparison details implementation and typical use cases for Soft Peripherals and High Performance Framework:

.. list-table::
   :header-rows: 1

>    * - Aspect

Reviewer: Aspect of what?

Author: added comparison

>      - Soft Peripherals

Reviewer: Is the name capitalized or not? Earlier in line 24 it isn't, here it is.

     - High Performance Framework
   * - Development
     - Developed by Nordic Semiconductor, optimized for performance and power consumption.

>      - Allows for the creation of specialized or custom peripherals.

Reviewer: Does this mean HPF is not developed by Nordic? Can we state this?

Reviewer: Also, I feel like this table is repeating information from the lines 24-28. Maybe remove the table and integrate the information into the lines there, so that we avoid duplication? Or the opposite?

   * - Use cases
     - When an application requires an additional or unavailable peripheral and a corresponding Soft Peripheral is available.
     - When developing a custom peripheral or when high-degree control is needed.

>    * - Suitability

Reviewer: ```suggestion
   * - Recommended usage
```

Reviewer: I actually feel "Use cases" and "Suitability" are quite overlapping. Can these be merged together?

.. list-table::
   :header-rows: 1

   * - Feature
     - Soft Peripherals
     - High Performance Framework
   * - Integration
     - Pre-compiled binary for FLPR, driver for the application core
     - Built from source in Zephyr build system
   * - Memory requirements
     - Fixed (~16 K)
     - Adjustable, depends on code and devicetree
   * - Application compatibility
     - Can be used on baremetal or Zephyr applications
     - Uses Zephyr without kernel on FLPR side, full Zephyr on APP side
   * - Integration complexity
     - Minor integration required
     - Samples provided for the start of development
   * - IPC mechanism
     - Register-based
     - ic(b)msg or mbox-based
   * - Compliance and testing
     - Protocol-compliant, verified by Nordic Semiconductor
     - Samples provided as-is, testing is done by the user
   * - Delivery form
     - Binary
     - Source code, modifiable and extendable
   * - API level
     - Exposed at hardware driver level
     - FLPR may handle parts of the software stack above hardware driver level

.. _nrf54l_hpf_softperi_comparison_creating_peripherals:

> Creation of software-based peripherals

Reviewer: I'm a bit confused given that "soft peripherals", "software-defined peripherals", and "software-based peripherals" are all used on one page. Are they all the same? ;)
